core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -3964,6 +3965,23 @@ private void handleSuccess(PersistJob job) {
               String ufsPath = resolution.getUri().toString();
               ufs.setOwner(tempUfsPath, inode.getOwner(), inode.getGroup());
               ufs.setMode(tempUfsPath, inode.getMode());
+
+              // Check if the size is the same to guard against a race condition where the Alluxio
+              // file is mutated in between the persist command and execution
+              if (ServerConfiguration.isSet(PropertyKey.MASTER_ASYNC_PERSIST_SIZE_VALIDATION)
+                  && ServerConfiguration.getBoolean(
+                      PropertyKey.MASTER_ASYNC_PERSIST_SIZE_VALIDATION)) {
+                UfsStatus ufsStatus = ufs.getStatus(tempUfsPath);
+                if (ufsStatus.isFile()) {
+                  UfsFileStatus status = (UfsFileStatus) ufsStatus;
+                  if (status.getContentLength() != inode.getLength()) {
+                    throw new IOException(String.format("%s size does not match. Alluxio expected"
+                        + "length: %d, UFS actual length: %d", tempUfsPath,
```

Comment:
(nit) add the alluxio file path and file id

Reply:
These will get logged further down in the catch clause.